### PR TITLE
[ReturnTypeExtension] add test case for narrow typed json_decode

### PR DESCRIPTION
**Pull request by TomasVotruba**
Port of https://github.com/phpstan/phpstan-nette/pull/89/files for `json_decode()`

Any feedback appreciated :+1: 

### How to test?

```bash
vendor/bin/phpunit tests/PHPStan/Analyser/NodeScopeResolverTest.php --filter "json"
```

### CODE REVIEW

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -99,4 +119,73 @@ private function isBitwiseOrWithJsonThrowOnError(Expr $expr, Scope $scope): bool
...
+		}
+
+		// prefer specific type
+		if (! $resolvedType instanceof MixedType) {
+			return $resolvedType;
+		}
+
+		// fallback type
+		if ($isForceArray) {
+			return new UnionType([
```
Comment (ondrejmirtes) [inline src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php:143]: As I already said in https://github.com/phpstan/phpstan-nette/pull/89#discussion_r795243535, we really can't use a union. Right now people using `json_decode` are "penalized" on level 9 thanks to `mixed`. With this precise union, they'd be penalized on level 7. That's not something we can do in a minor version, maybe ever.

The best we can do here I think is that we can return `MixedType` for case with `$assoc=false` and `MixedType minus stdClass` for `$assoc=true`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php (hunk 1 above)
Comment (TomasVotruba) [reply src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: I've update the fix :+1:

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -99,4 +109,51 @@ private function isBitwiseOrWithJsonThrowOnError(Expr $expr, Scope $scope): bool
...
+		if ($isForceArray) {
+			return new MixedType(true, new ObjectType(stdClass::class));
+		}
+
+		return new MixedType(true);
+	}
+
+	/**
+	 * Is "json_decode(..., true)"?
+	 * @param Arg[] $args
```
Comment (ondrejmirtes) [inline src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php:134]: build failure here

File: tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+assertType('1.5', $value);
+
+$value = json_decode('false', true);
+assertType('false', $value);
+
+$value = json_decode('{}', true);
+assertType('array{}', $value);
+
+$value = json_decode('[1, 2, 3]', true);
+assertType('array{1, 2, 3}', $value);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:23]: What's gonna happen in case of invalid json?

File: tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+$value = json_decode('true', true);
+assertType('true', $value);
+
+$value = json_decode('1', true);
+assertType('1', $value);
+
+$value = json_decode('1.5', true);
+assertType('1.5', $value);
+
+$value = json_decode('false', true);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:16]: Please also test second argument `false` and also please test `null` with `JSON_OBJECT_AS_ARRAY`: 

> When true, JSON objects will be returned as associative arrays; when false, JSON objects will be returned as objects. When null, JSON objects will be returned as associative arrays or objects depending on whether JSON_OBJECT_AS_ARRAY is set in the flags.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php (hunk 3 above)
Comment (TomasVotruba) [reply tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:23]: Looks like `null` is returned: https://3v4l.org/oPLja

File: tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php (hunk 4 above)
Comment (TomasVotruba) [reply tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:16]: I'm sure how the `JSON_OBJECT_AS_ARRAY` exactly works. I haven't use `json_decode` in last couple of years.

What is the exptected type? It seems the result with `JSON_OBJECT_AS_ARRAY` is same as 2nd parameter `true`
https://3v4l.org/ffeIM

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:16]: Yes, but we need to check that PHPStan understands it as well when called as `json_decode($input, null, JSON_OBJECT_AS_ARRAY)`. And also the other behaviour when called without it.

Comment (TomasVotruba) [reply tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:16]: I've added separate `json_object_as_array` to cover it.

Maybe one more test in case of combination of constants.

File: tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php (hunk 3 above)
Comment (TomasVotruba) [reply tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:23]: I've added test fixture for it :+1:

**Comment by TomasVotruba** [comment]
I'm working on the static errors...

**Comment by ondrejmirtes** [comment]
Feel free to open a new PR if you ever decide to work on this again 👍 Thanks.

**Comment by TomasVotruba** [comment]
It's ready from my side :+1: 

I'm actually waiting for feedback last 2 months: 

image

**Comment by ondrejmirtes** [comment]
It is not - there were conflicts and build failures.

File: composer.json [hunk 5]
```diff
@@ -6,6 +6,7 @@
 	],
 	"require": {
 		"php": "^8.0",
+		"ext-json": "*",
```
Comment (ondrejmirtes) [inline composer.json]: This isn't needed in composer.json, please remove it.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
and please fix the conflict by cleanly rebasing the branch on top of 1.5.x. I'll review it once the build is green. Thanks.

**Comment by TomasVotruba** [comment]
I'm on it :+1:

**Comment by TomasVotruba** [comment]
I've rebased. Tests, PHPStan and coding standards are passing localy.

Seems GitHub Actions are crashing now because cache is stuck: https://github.com/phpstan/phpstan-src/runs/5677735280?check_suite_focus=true

I'll check it today/tomorrow

File: tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php (hunk 4 above)
Comment (TomasVotruba) [reply tests/PHPStan/Analyser/data/json-decode/narrow_type_with_force_array.php:16]: This is the fixture file → https://github.com/phpstan/phpstan-src/pull/993/files#diff-7dcc5140f8f437277f737c9d311182df3bbe59afa02e59768dcccef72a514b1c

File: composer.json (hunk 5 above)
Comment (TomasVotruba) [reply composer.json]: How should I fix the missing extension failing CI then?

https://github.com/phpstan/phpstan-src/runs/5677868785?check_suite_focus=true

image

Comment (ondrejmirtes) [reply composer.json]: Put them in this file: https://github.com/phpstan/phpstan-src/blob/1.5.x/build/composer-require-checker.json

Comment (TomasVotruba) [reply composer.json]: Thx

**Comment by TomasVotruba** [comment]
CI is passing :+1:  
What is now missing from my side?

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php [hunk 6]
```diff
@@ -66,4 +94,85 @@ public function getTypeFromFunctionCall(
...
+
+	/**
+	 * Is "json_decode(..., true)"?
+	 */
+	private function isForceArrayWithoutStdClass(FuncCall $funcCall, Scope $scope): bool
+	{
+		$args = $funcCall->getArgs();
+
+		if (isset($args[1])) {
+			$secondArgValue = $this->resolveMaskValue($args[1]->value, $scope);
```
Comment (staabm) [inline src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: Not sure whether I understand this mask-resolving magic, but maybe it could re-use https://github.com/phpstan/phpstan-src/blob/1.5.x/src/Type/BitwiseFlagHelper.php instead?

Comment (ondrejmirtes) [reply src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: Yes, please use the new helper instead of ConstExprEvaluator.

Comment (TomasVotruba) [reply src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: I've tried it and it works for bit constant. On this line the `true` is also resolved, so it still need to resolve the value.

Comment (staabm) [reply src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: shouldn't it be someting like `if (BooleanType) handle-bool ELSE use-bitwiseFlagAnalyser` ?

Comment (TomasVotruba) [reply src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: Could you share more code, please? I'm not sure what exactly do you mean.

**Comment by herndlm** [comment]
I have a follow-up in mind and want to get this merged. Added 2 simplification commits via https://github.com/TomasVotruba/phpstan-src/pull/1 that deal also with the remaining discussion and should help moving this forward hopefully :)

I guess this PR should also be rebased and target 1.6.x instead? Pretty sure I broke a test case via a recent contribution ;)

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php [hunk 7]
```diff
@@ -51,8 +71,15 @@ public function getTypeFromFunctionCall(
 		Scope $scope,
 	): Type
 	{
+		// update type based on JSON_THROW_ON_ERROR
 		$argumentPosition = $this->argumentPositions[$functionReflection->getName()];
 		$defaultReturnType = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		// narrow type for json_decode()
```
Comment (herndlm) [inline src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: nitpick: I think there's no need for this comment, this block reads quite well IMO and is good to understand
```suggestion
```

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php [hunk 8]
```diff
@@ -51,8 +71,15 @@ public function getTypeFromFunctionCall(
 		Scope $scope,
 	): Type
 	{
+		// update type based on JSON_THROW_ON_ERROR
```
Comment (herndlm) [inline src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]:
```suggestion
```
this comment is actually not true any more, is it? the extension would be doing more then, I'd even argue to rename it in a follow-up :)

**Comment by herndlm** [comment]
Hi @TomasVotruba, I don't want to bother you too much, but do you mind taking a look at the PR I opened against your branch? Alternatively, if you don't mind, I can also open a new PR, trying to finish this. I'd keep your commits in of course :)

**Comment by TomasVotruba** [comment]
Hi @herndlm , thanks for the ping. Could you link the PR?
This is finished from my side, I'm waiting for feedback now :) 

Do you see anything missing @ondrejmirtes?

File: src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php (hunk 6 above)
Comment (ondrejmirtes) [reply src/Type/Php/JsonThrowOnErrorDynamicReturnTypeExtension.php]: I agree with @staabm, we don't need `constExprEvaluator` here. PHPStan is already a type inference engine by itself.

So `$scope->getType($args[1]->value)` will tell you whether it's `true` or `false`, otherwise you can use the `bitwiseFlagAnalyser`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also please rebase on top of 1.6.x, there isn't going to be another release of 1.5.x.

**Comment by herndlm** [comment]
> Hi @herndlm , thanks for the ping. Could you link the PR? This is finished from my side, I'm waiting for feedback now :)

@TomasVotruba https://github.com/TomasVotruba/phpstan-src/pull/1

I mentioned it in a comment a couple of days ago, but I should have just tagged you. thx! :)

**Comment by herndlm** [comment]
Tomas gave me access to his fork and I pushed my cleanups (after Tomas did the rebase). I didn't change too much on purpose. Let me know what is still needed, I'm taking care of getting this merged :)

**Comment by herndlm** [comment]
ah unfortunately we missed 1.6.0, but I hope we can still get this in 1.6.x? no pressure :D

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by TomasVotruba** [comment]
:+1: 

@herndlm Thank you for finishing :clap:

**Comment by herndlm** [comment]
Nice, thank you too. I was waiting for this getting merged, have a tiny follow-up idea :)
